index.js

```diff
@@ -109,6 +182,271 @@ export const immediate = x => Future ((rej, res) => {
+  const location = new URL (url);
+  return chain (lib => Future ((rej, res) => {
+    const req = lib.request (location, options, res);
+    pipeline (body, req, e => e && rej (e));
+    return () => {
+      req.abort ();
+    };
+  })) (getRequestModule (location.protocol));
+};
+
+//# messageMeta :: IncomingMessage -> MessageMeta
+//.
+//. Extracts the common metadata from an [IncomingMessage][].
+//.
+//. The `MessageMeta` type is defined as
+//. `{ code :: Number, message :: String, headers :: StrMap String }`:
+//.
+//. * `code`: The HTTP status code.
+//. * `message`: The HTTP status message.
+//. * `headers`: A StrMap of the HTTP headers.
+export const messageMeta = message => ({
+  code: message.statusCode,
+  message: message.statusMessage,
+  headers: message.headers,
+});
+
+//# bufferResponse :: Charset -> IncomingMessage -> Future Error { meta :: MessageMeta, body :: String }
+//.
+//. A version of [`buffer`](#buffer) specialized in [IncomingMessage][]s.
+//.
+//. Given a charset and an IncomingMessage, returns a Future with the
+//. [MessageMeta](#messageMeta) and the buffered, encoded, message body.
+export const bufferResponse = charset => message => {
+  const meta = messageMeta (message);
+  const eventualBody = mapRej (e => new Error ('Failed to buffer response: ' +
+                                               e.message))
+                              (bufferString (charset) (message));
+  return map (body => ({meta, body})) (eventualBody);
+};
+
+// defaultCharset :: String
+const defaultCharset = 'utf8';
+
+// defaultContentType :: String
+const defaultContentType = 'text/plain; charset=' + defaultCharset;
+
+// charsetRegex :: RegExp
+const charsetRegex = /\bcharset=([^;\s]+)/;
+
+//# autoBufferResponse :: IncomingMessage -> Future Error { meta :: MessageMeta, body :: String }
+//.
+//. Given an IncomingMessage, buffers and decodes the message body using the
+//. charset provided in the message headers. Falls back to UTF-8 if the
+//. charset was not provided.
+//.
+//. Returns a Future with the [MessageMeta](#messageMeta) and the buffered,
+//. encoded, message body.
+export const autoBufferResponse = message => {
+  const contentType = message.headers['content-type'] || defaultContentType;
+  const parsed = charsetRegex.exec (contentType);
+  const charset = parsed == null ? defaultCharset : parsed[1];
+  return bufferResponse (charset) (message);
+};
+
+//# retrieve :: Url -> StrMap String -> Future Error { meta :: MessageMeta, body :: String }
+//.
+//. A version of [`request`](#request) specialized in the `GET` method.
+//.
+//. Given a URL and a StrMap of request headers, returns a Future which
+//. makes a GET requests to the given resource, buffers its response and
+//. encodes it accordingly, falling back to UTF-8 encoding as described in
+//. [`autoBufferResponse`](#autoBufferResponse).
+//.
+//. ```js
+//. retrieve ('https://api.github.com/users/Avaq')
+//.          ({headers: {'User-Agent': 'Avaq'}})
```

Comment:
If I am not mistaken, the second argument should be an Object containing only the headers.

```js
retrieve ('https://api.github.com/users/Avaq')
         ({'User-Agent': 'Avaq'})
```